src/gc/collector.cpp

```diff
@@ -342,26 +416,121 @@ void markPhase() {
 #endif
 }
 
+void finalizationOrderingFirstPass(Box* obj) {
+    std::shared_ptr<FirstPhaseStack> stack = std::make_shared<FirstPhaseStack>();
+    GCVisitor visitor(stack);
+
+    stack->push(obj);
+    while (void* p = stack->pop()) {
+        visitByGCKind(p, visitor);
+    }
+}
+
+void finalizationOrderingSecondPass(Box* obj) {
+    std::shared_ptr<SecondPhaseStack> stack = std::make_shared<SecondPhaseStack>();
+    GCVisitor visitor(stack);
+
+    stack->push(obj);
+    while (void* p = stack->pop()) {
+        GCAllocation* al = GCAllocation::fromUserData(p);
+        assert(orderingState(al) != FinalizationState::UNREACHABLE);
+        visitByGCKind(p, visitor);
+    }
+}
+
+// Implementation of PyPy's finalization ordering algorithm:
+// http://pypy.readthedocs.org/en/latest/discussion/finalizer-order.html
+void finalizationOrderingPhase() {
+    // TODO: Replace with iterator or something that takes a lambda for memory efficiency?.
+    std::vector<Box*> objects_with_finalizers;
+    global_heap.getOrderedFinalizers(objects_with_finalizers);
```

Comment:
I replied to the original comment.

If we build `objects_with_finalizers` during allocation, `__del__` methods added after the fact will be a problem. I'm not sure if that'd be hard, or if it'd even be an edge case worth handling?